### PR TITLE
Prepare borgmatic change

### PR DESCRIPTION
**Pull request by randomn4me**
### Description

This is preparing an upcoming change in borgmatic to make repositories consisting of paths only unusable. This addresses issue #4580 .

This PR:

1. Is backwards compatible. Paths only in the config will be translated to the required `{ path = "path/to/repo"; }`
2. Paths can be given as
```
    {
      path = "path";
      label = "label";
    }
```
3. Paths can be given as just `{ path = "path/to/repo"; }`, however, this resolves to a config file, which contains `null` as label, which is forbidden by `borgmatic`.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

```
borgmatic-program-basic-configuration: OK
borgmatic-program-exclude-hm-symlinks: OK
borgmatic-program-exclude-hm-symlinks-nothing-else: OK
borgmatic-program-include-hm-symlinks: OK
borgmatic-service-basic-configuration: OK
```

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

**Comment by rycee** [comment]
Thanks for the contribution! Looks OK overall, I would suggest adding an [`apply`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.options.mkOption) attribute to the option to convert string values into attribute set ones. If you search the HM repo for "apply =" you should find a number of examples.

To complete the checklist you need to squash the commits into a single one having a commit message matching the given format. No need to do immediately, you can do this once the PR is close to being merged. Also you need to add a test. Can have a simple test with three repositories, one a plain string, one an attribute set with only a path, and one an attribute set with both path and label.

Maintainer CC @DamienCassou

File: modules/programs/borgmatic.nix
Comment (DamienCassou) [inline modules/programs/borgmatic.nix:17]:
```suggestion
    map (repo: if builtins.isString repo then { path = repo; } else removeNullValues repo) repos;
```

File: modules/programs/borgmatic.nix
Comment (DamienCassou) [inline modules/programs/borgmatic.nix:44]: I would give a little more info here. Maybe something like

```suggestion
        description = "Short text describing the repository. Can be used with the `--repository` flag to select a repository";
```

File: modules/programs/borgmatic.nix
Comment (DamienCassou) [inline modules/programs/borgmatic.nix:80]: "Paths to key/value pairs" doesn't mean anything to me. Maybe something like:

```suggestion
          description = "List of local or remote repositories with paths and optional labels.";
```

**Review by DamienCassou — CHANGES_REQUESTED** [review]
Great job, thank you!

**Comment by randomn4me** [comment]
@rycee I adjusted the testcases (3 now: 1 str, 1 `{ path="path"}`, 1 `{path="path", label="label"}`) and added @DamienCassou 's commits. If it is now finished I need to find out how to squash and then finish this PR.

File: modules/programs/borgmatic.nix
Comment (DamienCassou) [inline modules/programs/borgmatic.nix:84]: @rycee suggested to use `apply` here instead of deferring cleaning until much later in the file:

```suggestion
          type = types.listOf (types.either types.str repositoryOption);
          apply = cleanRepositories;
```

File: modules/programs/borgmatic.nix
Comment (DamienCassou) [inline modules/programs/borgmatic.nix:193]: this should no be necessary if using `apply` as suggested

File: tests/modules/programs/borgmatic/basic-configuration.nix
Comment (DamienCassou) [inline tests/modules/programs/borgmatic/basic-configuration.nix:22]: assertions much be changed accordingly. You should replace:

```
    expectations[repositories[0]]="${
      builtins.elemAt backups.main.location.repositories 0
    }"
    expectations[repositories[1]]="${
      builtins.elemAt backups.main.location.repositories 1
    }"
```

with:

```
    expectations[repositories[0].path]="${
      (builtins.elemAt backups.main.location.repositories 0).path
    }"
    expectations[repositories[1].path]="${
      (builtins.elemAt backups.main.location.repositories 1).path
    }"
    expectations[repositories[2].path]="${
      (builtins.elemAt backups.main.location.repositories 2).path
    }"
    expectations[repositories[2].label]="${
      (builtins.elemAt backups.main.location.repositories 2).label
    }"
```

**Review by DamienCassou — COMMENTED** [review]
Thank you very much. I think there are 2 more little changes to do and we will be good to merge.

File: modules/programs/borgmatic.nix
Comment (randomn4me) [reply modules/programs/borgmatic.nix:193]: I'll remove it

**Comment by randomn4me** [comment]
It should all be done now. If there is more that needs to be done before a PR, just ping me.

**Comment by DamienCassou** [comment]
The result is perfect!

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Pull request merged**
